OctoAwesome/OctoAwesome/Components/ComponentList.cs

```diff
@@ -9,202 +8,360 @@
+                existing.Add(component);
+
+        }
+        else
+        {
+            componentsByType[type] = new() { component };
         }
 
-        /// <summary>
-        /// Returns the Component of the given Type or null
-        /// </summary>
-        /// <typeparam name="V">Component Type</typeparam>
-        /// <returns>Component</returns>
-        public V? GetComponent<V>()
+        flatComponents.Add(component);
+        insertValidator?.Invoke(component);
+        onInserter?.Invoke(component);
+    }
+
+
+    /// <summary>
+    /// Checks whether the component of <typeparamref name="V"/> is present in the internal dictionary as a key.
+    /// </summary>
+    /// <typeparam name="V">The type to search in the internal dictionary</typeparam>
+    /// <returns>
+    /// <list type="bullet">
+    ///     <item><see langword="true"/> if the component was found</item>
+    ///     <item><see langword="false"/> if the component was not found</item>
+    /// </list>
+    /// </returns>
+    public bool Contains<V>()
+    {
+        var type = typeof(V);
+        foreach (var x in TypeKeys)
         {
-            if (components.TryGetValue(typeof(V), out var result))
-                return (V)(object)result;
+            if (type.IsAssignableFrom(x))
+                return true;
+        }
+        return false;
+    }
 
-            return default;
+    /// <summary>
+    /// Tries to return the Component of the given Type or null
+    /// </summary>
+    /// <typeparam name="V">Component Type</typeparam>
+    /// <returns>True if the component was found, false otherwise</returns>
+    public bool TryGet<V>([MaybeNullWhen(false)] out V component) where V : T
+    {
+        var contains = componentsByType.TryGetValue(typeof(V), out var result);
+        if (!contains || result is null || result.Count < 1)
+        {
+            component = default;
+            return false;
         }
 
-        /// <summary>
-        /// Removes the Component of the given type.
-        /// </summary>
-        /// <typeparam name="V">The type of the component to remove.</typeparam>
-        /// <returns>A value indicating whether the remove was successful or not.</returns>
-        public bool RemoveComponent<V>() where V : T
+        component = GenericCaster<T, V>.Cast(result[0]);
+        return contains;
+    }
+
+    /// <summary>
+    /// Returns the first Component of the given Type or null
+    /// </summary>
+    /// <typeparam name="V">Component Type</typeparam>
+    /// <returns>Component</returns>
+    public V? Get<V>()
+    {
+        if (componentsByType.TryGetValue(typeof(V), out var result) && result.Count > 0)
+            return GenericCaster<T, V>.Cast(result[0]);
+
+        return default;
+    }
+
+    /// <summary>
+    /// Returns a list of the Component with the given Type or null
+    /// </summary>
+    /// <typeparam name="V">Component Type</typeparam>
+    /// <returns>Component</returns>
```

Comment:
```suggestion
    /// Gets a list of components of the given type.
    /// </summary>
    /// <typeparam name="V">The component type to search for.</typeparam>
    /// <returns>A list of components if any was found; otherwise <c>null</c>.</returns>
```